File: src/main/java/com/fusi24/pangreksa/base/ui/view/MainErrorHandler.java
package com.fusi24.pangreksa.base.ui.view;

import com.fusi24.pangreksa.base.ui.notification.AppNotification;
import com.vaadin.flow.component.Component;
import com.vaadin.flow.server.VaadinServiceInitListener;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;
import org.springframework.context.annotation.Bean;
import org.springframework.context.annotation.Configuration;

@Configuration
class MainErrorHandler {

    private static final Logger log = LoggerFactory.getLogger(MainErrorHandler.class);

    @Bean
    public VaadinServiceInitListener errorHandlerInitializer() {
        return (event) -> event.getSource().addSessionInitListener(
                sessionInitEvent -> sessionInitEvent.getSession().setErrorHandler(errorEvent -> {

                    log.error("Terjadi kesalahan yang tidak terduga", errorEvent.getThrowable());

                    errorEvent.getComponent()
                            .flatMap(Component::getUI)
                            .ifPresent(ui -> ui.access(() ->
                                    AppNotification.error(
                                            "Terjadi kesalahan sistem. Silakan coba beberapa saat lagi."
                                    )
                            ));
                }));
    }
}

File: src/main/kotlin/com/fusi24/pangreksa/web/base/ui/view/MainErrorHandler.kt
package com.fusi24.pangreksa.web.base.ui.view

import com.vaadin.flow.component.Component
import com.vaadin.flow.component.notification.Notification
import com.vaadin.flow.component.notification.NotificationVariant
import com.vaadin.flow.server.ErrorEvent
import com.vaadin.flow.server.ErrorHandler
import com.vaadin.flow.server.VaadinServiceInitListener
import org.slf4j.Logger
import org.slf4j.LoggerFactory
import org.springframework.context.annotation.Bean
import org.springframework.context.annotation.Configuration

@Configuration
open class MainErrorHandler {

    companion object {
        private val log: Logger = LoggerFactory.getLogger(MainErrorHandler::class.java)
    }

    @Bean
    open fun errorHandlerInitializer(): VaadinServiceInitListener {
        return VaadinServiceInitListener { event ->
            event.source.addSessionInitListener { sessionInitEvent ->
                sessionInitEvent.session.errorHandler = { errorEvent: ErrorEvent ->
                    log.error("An unexpected error occurred", errorEvent.throwable)
                    errorEvent.component.flatMap { component: Component? ->
                                    component?.ui
                                }.ifPresent { ui ->
                                    val notification = Notification(
                                        "An unexpected error has occurred. Please try again later."
                                    )
                                    notification.addThemeVariants(NotificationVariant.LUMO_ERROR)
                                    notification.position = Notification.Position.TOP_CENTER
                                    notification.duration = 3000
                                    ui.access { notification.open() }
                                }
                } as ErrorHandler
            }
        }
    }
}